src/btree/row_srch.c

```diff
@@ -765,6 +765,15 @@ __wt_row_random_leaf(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt)
 	cbt->ins_head = ins_head;
 	cbt->compare = 0;
 
+	/*
+	 * MongoDB $sample can be significantly slower than expected if the
+	 * collection is newly created and the content is in a in a single
+	 * skiplist. If searching a relatively large skiplist, schedule it
+	 * for eviction to split up the skiplist.
+	 */
+	if (page->memory_footprint > 5 * WT_MEGABYTE)
```

Comment:
The performance problem appears in pages with big skiplists but not yet large enough to be forcibly evicted, that is, pages with a big footprint but less than `memory_page_max`.

I considered using `memory_page_max`, but decided `memory_page_max / 2` was more likely to go wrong than `5MB`: this really is "what memory footprint implies a skiplist too large to walk 100K times in the next couple of seconds?"